### PR TITLE
Add getcustomtx RPC call

### PR DESCRIPTION
**Pull request by Bushstar**
Adds new RPC call getcustomtx to return details on DeFiChain custom transactions. Can only reliable return information for transactions if they are wallet transaction, in the mempool, a blockhash is provided or the transaction index has been enabled. The user is informed to provide a blockhash or enable the transaction index if a transaction cannot be found and neither of these are already available.

The "valid" key will indicate if the custom transaction is valid. A transaction can reliably be determined to be valid if it is a wallet transaction, txindex is enabled or a block hash is provided, otherwise results from the transaction parsing functions will be used which cannot will has to be partial validation as the exact state at the time of the original transaction cannot be replicated.

#### Transaction not found example:
```
./src/defi-cli -regtest getcustomtx 8798b2da967d10a6cf82a3a4e5cb1ac3f1712b356f194cc2f5b94aeebbaeae55
error code: -5
error message:
No such mempool or wallet transaction. Use -txindex or provide a block hash.
```

#### createtoken
```
{
  "type": "CreateToken",
  "valid": true,
  "results": {
    "creationTx": "f5c33800d28706e92d68ae946b5f15e9d2a4a4b95f92e97b772c1f5ef373df4a",
    "name": "gold",
    "symbol": "GOLD",
    "isDAT": false,
    "mintable": true,
    "tradeable": true,
    "finalized": false
  },
  "blockhash": "13e8614d881630df9a5d3484714886cffbe9351f1a96c3438a47d4cc937e8448",
  "block height": 102,
  "blocktime": 1608561403,
  "confirmations": 18
}

```
#### minttokens
```
{
  "type": "MintToken",
  "valid": true,
  "results": {
    "128": 300.00000000
  },
  "blockhash": "278bbf0c23e3c8ff607788bad16ae6511d7924f422895aada08df76e73e5a633",
  "block height": 103,
  "blocktime": 1608561404,
  "confirmations": 17
}
```
#### updatetoken
```
{
  "type": "UpdateTokenAny",
  "valid": true,
  "results": {
    "name": "silver",
    "symbol": "SILVER",
    "isDAT": false,
    "mintable": false,
    "tradeable": false,
    "finalized": false
  },
  "blockhash": "0204de7a4255c78d764f5d891acb13af2bffce0794d2ff3097c00aa14fcba49e",
  "block height": 104,
  "blocktime": 1608561405,
  "confirmations": 16
}
```
#### createmasternode
```
{
  "type": "CreateMasternode",
  "valid": true,
  "results": {
    "masternodeoperator": "mj8Uyn8m87eTRDJnuQQ3SWwEP7Q3JkJk9V"
  },
  "blockhash": "d6d4056fe5b10f5282efbff3f4086c77ce2abd3a586c28cc2f4bf335d071cd8b",
  "block height": 105,
  "blocktime": 1608561406,
  "confirmations": 15
}
```
#### resignmasternode
```
{
  "type": "ResignMasternode",
  "valid": true,
  "results": {
    "id": "b0adae4649d8154f86d1e2d55205ad8cab2198534a2c330893d8343f7619ed78"
  },
  "blockhash": "1306fef1d309c013b111fc5b14f26bd3d8b8c5a5cc3353f6462595322cf2500c",
  "block height": 106,
  "blocktime": 1608561407,
  "confirmations": 14
}
```
#### createpoolpair
```
{
  "type": "CreatePoolPair",
  "valid": true,
  "results": {
    "creationTx": "819cff914f7fb0a140c6d9e9383844f8a5f738ff0dff528bd4052309e78de087",
    "name": "silver-gold",
    "symbol": "SILVGOLD",
    "tokenA": "silver",
    "tokenB": "gold",
    "commission": 0.00100000,
    "status": true,
    "ownerAddress": "76a91489ad3518e7c168b962a66b3d2c4f55f7aeda906a88ac",
    "isDAT": true,
    "mineable": false,
    "tradeable": true,
    "finalized": true
  },
  "blockhash": "24ae14b916b802dda8fed3795a45823e93e5c7f489e1b751c4f39a1dff029ed5",
  "block height": 109,
  "blocktime": 1608561410,
  "confirmations": 11
}
```
#### accounttoaccount
```
{
  "type": "AccountToAccount",
  "valid": true,
  "results": {
    "from": "76a9142ea1d7e748db26b030f2c80d42e25e0eb829df5f88ac",
    "to": {
      "76a91400344877aaa15ff94c99114e7f728fdabab542fe88ac": "100.00000000@129"
    }
  },
  "blockhash": "8b562b960878e9764cb53b18a67eaa69b2f359a36f26f33210b87e0d469b92fa",
  "block height": 110,
  "blocktime": 1608561411,
  "confirmations": 10
}
```
#### addpoolliquidity
```
{
  "type": "AddPoolLiquidity",
  "valid": true,
  "results": {
    "128": 100.00000000,
    "129": 100.00000000,
    "shareaddress": "76a914745668169f594f83dc952bf48b6611486fa8027388ac"
  },
  "blockhash": "38bd98e95633ca97cbea322c0b892d78e2c534ed03869f7c18ae45e400ad5b8c",
  "block height": 111,
  "blocktime": 1608561412,
  "confirmations": 9
}
```
#### poolswap
```
{
  "type": "PoolSwap",
  "valid": true,
  "results": {
    "fromAddress": "76a91400344877aaa15ff94c99114e7f728fdabab542fe88ac",
    "fromToken": "128",
    "fromAmount": 1.00000000,
    "toAddress": "76a9142ea1d7e748db26b030f2c80d42e25e0eb829df5f88ac",
    "toToken": "129",
    "maxPrice": 2.00000000
  },
  "blockhash": "3ce0cb3f667bb238c870a65c9d12d6b030b4daa03a289ff384d752db67cab63b",
  "block height": 112,
  "blocktime": 1608561413,
  "confirmations": 8
}
```
#### removepoolliquidity
```
{
  "type": "RemovePoolLiquidity",
  "valid": true,
  "results": {
    "from": "76a914745668169f594f83dc952bf48b6611486fa8027388ac",
    "amount": "25.00000000@1"
  },
  "blockhash": "04b0194a606ff09583f915283bd8440b4d61e5e7330bdc07bf772994405fb4c5",
  "block height": 113,
  "blocktime": 1608561414,
  "confirmations": 7
}
```
#### updatepoolpair
```
{
  "type": "UpdatePoolPair",
  "valid": true,
  "results": {
    "commission": 0.10000000,
    "status": false,
    "ownerAddress": "76a91489ad3518e7c168b962a66b3d2c4f55f7aeda906a88ac"
  },
  "blockhash": "f767be33e1b29add600af25bc87b2549ce65f821281dc987e67fe5bcb203a2ff",
  "block height": 114,
  "blocktime": 1608561415,
  "confirmations": 6
}
```
#### utxostoaccount
```
{
  "type": "UtxosToAccount",
  "valid": true,
  "results": {
    "76a91400344877aaa15ff94c99114e7f728fdabab542fe88ac": "1.00000000@0"
  },
  "blockhash": "98d44483b5220d584336b7ee103e04ba184f2fdbc9972bd5f5153b02dd281bcc",
  "block height": 115,
  "blocktime": 1608561416,
  "confirmations": 5
}
```
#### accounttoutxos
```
{
  "type": "AccountToUtxos",
  "valid": true,
  "results": {
    "from": "76a91400344877aaa15ff94c99114e7f728fdabab542fe88ac",
    "to": {
      "76a9142ea1d7e748db26b030f2c80d42e25e0eb829df5f88ac": "1.00000000@0"
    }
  },
  "blockhash": "06700b21ed214c2a262de07cfd4a837e0b95cf78ed48bc9579e04746b6781578",
  "block height": 116,
  "blocktime": 1608561417,
  "confirmations": 4
}
```
#### sendtokenstoaddress
```
{
  "type": "AnyAccountsToAccounts",
  "valid": true,
  "results": {
    "from": {
      "76a91400344877aaa15ff94c99114e7f728fdabab542fe88ac": "1.00000000@0"
    },
    "to": {
      "76a9142ea1d7e748db26b030f2c80d42e25e0eb829df5f88ac": "1.00000000@0"
    }
  },
  "blockhash": "c61d2c8b8040224f462ea38a0f34a7fe1b78189f5319493490e6774bb6beefb2",
  "block height": 118,
  "blocktime": 1608561419,
  "confirmations": 2
}
```
#### setgov
```
{
  "type": "SetGovVariable",
  "valid": true,
  "results": {
    "LP_DAILY_DFI_REWARD": 35.00000000
  },
  "blockhash": "6cdc48f74f47c3adce9bfb580ebd0c7b6956cf06ca147b8efff35573c2537250",
  "block height": 119,
  "blocktime": 1608561420,
  "confirmations": 1
}
```

### CODE REVIEW

**Review by monstrobishi — APPROVED** [review]

File: src/masternodes/mn_rpc.cpp [hunk 1]
```diff
@@ -3754,6 +3755,213 @@ UniValue sendtokenstoaddress(const JSONRPCRequest& request) {
...
+        },
+    }.Check(request);
+
+    const uint256 hash(ParseHashV(request.params[0], "txid"));
+
+    CTransactionRef tx;
+    CBlockIndex* blockindex{nullptr};
+    uint256 hashBlock;
+    std::vector<unsigned char> metadata;
+    CustomTxType guess;
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:3793]: Can move the variable definition to the place it is been used?

File: src/masternodes/mn_checks.cpp [hunk 2]
```diff
@@ -1104,6 +1258,13 @@ Res ApplySetGovernanceTx(CCustomCSView &mnview, const CCoinsViewCache &coins, co
         if (!add.ok)
             return Res::Err("%s '%s': %s", base, name, add.msg);
     }
+
+    if (rpcInfo) {
+        for (auto& name : names) {
```
Comment (ShengguangXiao) [inline src/masternodes/mn_checks.cpp:1263]: Add const before auto

File: src/masternodes/mn_checks.cpp [hunk 3]
```diff
@@ -841,6 +939,23 @@ Res ApplyAnyAccountsToAccountsTx(CCustomCSView & mnview, CCoinsViewCache const &
...
+    // Return here to avoid balance errors below.
+    if (rpcInfo) {
+        UniValue source(UniValue::VOBJ);
+        for (auto& from : msg.from) {
+            source.pushKV(from.first.GetHex(), from.second.ToString());
+        }
+        rpcInfo->pushKV("from", source);
+
+        UniValue dest(UniValue::VOBJ);
+        for (auto& to : msg.to) {
```
Comment (ShengguangXiao) [inline src/masternodes/mn_checks.cpp:951]: Add const before auto

**Review by ShengguangXiao — CHANGES_REQUESTED** [review]

**Comment by monstrobishi** [comment]
@Bushstar there are a lot of changes in consensus areas, it looks pretty safe to me (you are just passing a null pointer each time), but are you absolutely certain this won't affect consensus?

File: src/masternodes/mn_rpc.cpp [hunk 4]
```diff
@@ -3768,6 +3769,213 @@ UniValue sendtokenstoaddress(const JSONRPCRequest& request) {
...
+            hashBlock = it->second.hashBlock;
+        }
+    }
+
+    CBlockIndex* blockindex{nullptr};
+
+    // No wallet or not a wallet TX, try mempool, txindex and a block if hash provided
+    if (!pwallet || !tx)
+    {
+        if (!request.params[1].isNull()) {
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:3822]: Should check the request.params size before accessing the element.

File: src/masternodes/mn_rpc.cpp [hunk 5]
```diff
@@ -3768,6 +3769,213 @@ UniValue sendtokenstoaddress(const JSONRPCRequest& request) {
...
+        result.pushKV("results", txResults);
+    }
+
+    if (!hashBlock.IsNull()) {
+        LOCK(cs_main);
+
+        result.pushKV("blockhash", hashBlock.GetHex());
+        if (blockindex) {
+            result.pushKV("block height", blockindex->nHeight);
+            result.pushKV("blocktime", blockindex->GetBlockTime());
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:3968]: I convention we are using "blockHeight" and "blockTime" as the key.

File: src/masternodes/mn_rpc.cpp [hunk 6]
```diff
@@ -3768,6 +3769,213 @@ UniValue sendtokenstoaddress(const JSONRPCRequest& request) {
...
+        // Skip coinbase TXs except for genesis block
+        if ((tx->IsCoinBase() && nHeight > 0)) {
+            return "Coinbase transaction. Not a custom transaction.";
+        }
+
+        std::vector<unsigned char> metadata;
+        guess = GuessCustomTxType(*tx, metadata);
+        CCustomCSView mnview_dummy(*pcustomcsview);
+
+        switch (guess)
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:3888]: Can we move this switch section to a seperate function, this function is quite long.

**Comment by Bushstar** [comment]
> @Bushstar there are a lot of changes in consensus areas, it looks pretty safe to me (you are just passing a null pointer each time), but are you absolutely certain this won't affect consensus?

This does not affect consensus.

File: src/masternodes/mn_rpc.cpp (hunk 4 above)
Comment (Bushstar) [reply src/masternodes/mn_rpc.cpp:3822]: No need, this is typical behaviour across Bitcoin codebase to check for a defined optional argument.

**Pull request merged**